Odometry

SyrianSpock opened the pull request:
Ready to be tested on the robot.

SyrianSpock committed:
Add readme file

SyrianSpock committed:
Setup infrastructure

SyrianSpock committed:
Correct testing setup & add script to automatise

SyrianSpock committed:
Basic functions & structure for single wheel manipulation

Missing: handling of transmission ratio

SyrianSpock committed:
Minor cleanups

SyrianSpock committed:
Implement constant acceleration motion model on a wheel

Able to predict current acceleration & velocity of a wheel at any time given the three previous samples

SyrianSpock committed:
Implement pose estimation w/ constant acceleration motion model

SyrianSpock committed:
Minor refactoring of pose estimation code

SyrianSpock committed:
Correcting pose estimation

No need to recompute an interpolation to apply the constance acceleration
motion model since this is already done for each wheel.
Tests are still incomplete

Stapelzeiger committed:
odometry formulas

SyrianSpock committed:
Start refactoring the whole implementation to make it simpler and more accurate

Redefine function call tree
Redefine structure used for computation
Finished refactoring encoder part
Todo: redo wheel & base part

SyrianSpock committed:
Finish refactoring implementation

Correct wheel update & prediction
Correct base pose/velocity estimation

SyrianSpock committed:
Complete odometry formulas

SyrianSpock committed:
Correct wheel prediction / interpolation

Fix computation of parabola parameters

SyrianSpock committed:
Fix tests and refactor

SyrianSpock committed:
Update documentation of the wheel prediction/interpolation

Remove old Euler style acceleration estimation
Add the parabola fitting equations

SyrianSpock committed:
Move formulas sheet to doc folder

SyrianSpock committed:
Clean up doc folder

Stapelzeiger commented on odometry.c:
```diff
@@ -0,0 +1,190 @@
+	int16_t delta;
+	uint16_t prediction;
+	prediction = wheel_predict(wheel, time_now) - wheel->samples[2].value;
+	delta = wheel->delta_pos_accumulator + prediction;
```
`delta = prediction - accumulator`, and it would be more logical if accumulator were uint16_t. (Although I think it should would work with int16_t since it is converted to uint16_t by the type promotion rules)

SyrianSpock replied:
I messed up the types, prediction should be a signed integer. Same for the accumulator, otherwise we can't handle the negative case.

Stapelzeiger replied:
I would prefer it to be unsigned, because singed overflow is not defined.

Stapelzeiger commented on odometry.c:
```diff
@@ -0,0 +1,190 @@
+		dy1 = (int16_t) (wheel->samples[1].value - wheel->samples[0].value);
+		dy2 = (int16_t) (wheel->samples[2].value - wheel->samples[0].value);
+		tau1 = dy1 / (- dt1 * dt1 + dt1 * dt2);
+		tau2 = dy2 / (- dt2 * dt2 + dt1 * dt2);
```
this will result in a division by zero for the first sample (all timestamps are equal, deltas are zero)

SyrianSpock replied:
Fixed in 3db53f5

Stapelzeiger commented on odometry.h:
```diff
@@ -0,0 +1,83 @@
+		odometry_wheel_t *wheel,
+		const timestamp_t time_now);
+
+float encoder_time_derivative(
```
this function is not used, and since you already provide speed information of the base, you don't need it.

SyrianSpock replied:
Yup, forgot to delete it.

Stapelzeiger commented on odometry.c:
```diff
@@ -0,0 +1,190 @@
+	}
+
+	robot->pose.x += delta_fwd * cos(robot->pose.theta + 0.5f * delta_rot);
+	robot->pose.y += delta_fwd * sin(robot->pose.theta + 0.5f * delta_rot);
```
it might be useful to cache the sin/cos since this would improve speed quite a lot.

SyrianSpock replied:
Sure.

antoinealb replied:
Actually you might want to use the `cost` function to avoid type promotion to double.

Stapelzeiger commented:
It looks mostly good.
A few additional points:
- use spaces for indentation
- prefix all exported functions with the module name
- provide getter functions or declare a struct instead of a typedef
- a lot of your functions should not be in the header. (you can either make them static and include the .c file for unit tests or declare the prototypes in the unit test.)

Stapelzeiger commented:
also maybe we should add a note that the odometry integration in main.tex is not what we actually implemented.

antoinealb commented:
Or better, update `main.tex` to reflect what is really implemented.

SyrianSpock committed:
Fix indentation using only spaces

SyrianSpock committed:
Correct computation of delta of the wheel

SyrianSpock committed:
Remove unused function

SyrianSpock committed:
Set internal functions to the module as static

SyrianSpock committed:
Avoid division by zero when fitting a parabola to the wheel's motion

SyrianSpock committed:
Refactor the code

Add module name as prefix to the functions
Provide getter functions for pose/velocity of the base
Cache cos/sin computed to improve speed of execution

SyrianSpock committed:
Update main documentation

Integrate formulas in main doc file and remove separate formula file
Reorganise main doc file
Update conventions

SyrianSpock commented:
Alright, I think I covered everything. 
I merged `formulas.tex` with `main.tex`, so that we have one file that documents everything.

SyrianSpock committed:
Use float version of cos/sin to avoid type promotion to double
